### PR TITLE
Support for Windowless/Offscreen Popups

### PR DESCRIPTION
**Pull request by jamespearce2006**
Initial solution for #869.

This PR isn't production-ready, but rather a starting point for a discussion. I'm new to CefSharp, so I've likely made some naive, or wrong-headed, architectural assumptions.

I've updated the WPF sample app (and struggled getting the change to fit the MVVM pattern).
To see the results, run the Wpf sample app and create a new tab. The new tab defaults to a w3schools try-it session, where you can open and close a window via Javascript.

### CODE REVIEW

**Comment by amaitland** [comment]
Can you fix the formatting, spaces instead of tabs, it's hard to read. If you look at the `Files Changed` tab above you'll see what I mean.

File: CefSharp.Core/ManagedCefBrowserAdapter.h [hunk 1]
```diff
@@ -70,7 +70,12 @@ namespace CefSharp
             DisposableResource::DoDispose(isDisposing);
         };
 
+	internal:
+		MCefRefPtr<ClientAdapter> GetClientAdapter();
+
     public:
+
```
Comment (amaitland) [inline CefSharp.Core/ManagedCefBrowserAdapter.h]: Remove extra white space, keep changes to a minimum.

File: CefSharp.Wpf.Example/LifespanHandler.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+using System.Threading.Tasks;
+using System.Windows;
+using System.Windows.Threading;
+
+namespace CefSharp.Wpf.Example
+{
+    public class LifespanHandler : ILifeSpanHandler
+    {
+
+        MainWindow _owner;
```
Comment (amaitland) [inline CefSharp.Wpf.Example/LifespanHandler.cs]: Only prefix with `_` in `c++` for consistency.

File: CefSharp.Wpf.Example/ViewModels/BrowserTabViewModel.cs [hunk 3]
```diff
@@ -14,6 +14,7 @@ namespace CefSharp.Wpf.Example.ViewModels
 {
     public class BrowserTabViewModel : ViewModelBase
     {
+
```
Comment (amaitland) [inline CefSharp.Wpf.Example/ViewModels/BrowserTabViewModel.cs]: Remove whitespace changes, specially as no changes were made to this class.

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml [hunk 4]
```diff
@@ -168,12 +168,15 @@
                     </Rectangle.Fill>
                 </Rectangle>
 
+                    <ContentControl Content="{Binding WebBrowser}"
+                                Opacity="{Binding ElementName=opacitySlider, Path=Value}" >
+                    <!--
                 <cefSharp:ChromiumWebBrowser x:Name="browser"
```
Comment (amaitland) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml]: Whilst your way of creating the `ChromiumWebBrowser` instance will obviously work, I think the extra level of indirection is not great for an example application.

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 5]
```diff
@@ -459,11 +459,12 @@ private static void OnAddressChanged(DependencyObject sender, DependencyProperty
 
         protected virtual void OnAddressChanged(string oldValue, string newValue)
         {
-            if (ignoreUriChange)
+            if (ignoreUriChange || newValue == null)
             {
                 return;
             }
 
+            //CreateOffscreenBrowserWhenActualSizeChanged();
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Remove this.

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 6]
```diff
@@ -459,11 +459,12 @@ private static void OnAddressChanged(DependencyObject sender, DependencyProperty
 
         protected virtual void OnAddressChanged(string oldValue, string newValue)
         {
-            if (ignoreUriChange)
+            if (ignoreUriChange || newValue == null)
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Slight change in behavior, whilst logical, will this affect existing users? To be honest I can't remember what passing `null` does, so would have to look further.

File: CefSharp/ILifeSpanHandler.cs [hunk 7]
```diff
@@ -17,8 +17,10 @@ public interface ILifeSpanHandler
         /// <param name="width">width</param>
         /// <param name="height">height</param>
         /// <param name="noJavascriptAccess">value indicates whether the new browser window should be scriptable
+        /// <param name="newBrowser">A newly created browser that will host the popup</param>
         /// and in the same process as the source browser.</param>
         /// <returns>To cancel creation of the popup window return true otherwise return false.</returns>
+
```
Comment (amaitland) [inline CefSharp/ILifeSpanHandler.cs]: Remove extra lines

File: CefSharp/ILifeSpanHandler.cs [hunk 8]
```diff
@@ -34,7 +36,13 @@ public interface ILifeSpanHandler
         /// the new browser window should be scriptable and in the same process as the
         /// source browser.
         /// </remarks>
-        bool OnBeforePopup(IWebBrowser browser, IFrame frame, string targetUrl, ref int x, ref int y, ref int width, ref int height, ref bool noJavascriptAccess);
+        bool OnBeforePopup(IWebBrowser browser, IFrame frame, string targetUrl, ref int x, ref int y, ref int width, ref int height, ref bool noJavascriptAccess, ref IWebBrowser newBrowser);
+
+        /// <summary>
+        /// Called after a new browser is created.
+        /// </summary>
+        /// <param name="browser"></param>
```
Comment (amaitland) [inline CefSharp/ILifeSpanHandler.cs]: Add comment (just copy and paste)

File: CefSharp.Wpf.Example/MainWindow.xaml [hunk 9]
```diff
@@ -2,8 +2,20 @@
         xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
         xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
         xmlns:controls="clr-namespace:CefSharp.Wpf.Example.Controls"
+        xmlns:cefSharp="clr-namespace:CefSharp.Wpf;assembly=CefSharp.Wpf" xmlns:local="clr-namespace:CefSharp.Wpf.Example.ViewModels"
```
Comment (amaitland) [inline CefSharp.Wpf.Example/MainWindow.xaml]: Wrap lines.

File: CefSharp.Core/ManagedCefBrowserAdapter.cpp [hunk 10]
```diff
@@ -20,6 +20,7 @@ void ManagedCefBrowserAdapter::CreateOffscreenBrowser(IntPtr windowHandle, Brows
     }
 }
 
+
```
Comment (amaitland) [inline CefSharp.Core/ManagedCefBrowserAdapter.cpp]: Remove extra line

File: CefSharp.Core/Internals/ClientAdapter.h [hunk 11]
```diff
@@ -62,6 +62,8 @@ namespace CefSharp
             void CloseDevTools();
             void CloseAllPopups(bool forceClose);
 
+			virtual bool IsOffscreen() { return false; }
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.h]: Should be a property.

File: CefSharp.Core/Internals/ClientAdapter.cpp [hunk 12]
```diff
@@ -72,19 +72,42 @@ namespace CefSharp
...
+
+			return returnVal;
+
         }
 
         void ClientAdapter::OnAfterCreated(CefRefPtr<CefBrowser> browser)
-        {
-            if (browser->IsPopup())
+        {			
+            if (browser->IsPopup() && !IsOffscreen())
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Without running the code the appearance of `IsOffscreen()` seems rather odd. Which `ChromiumWebBrowser` instances are the events been propagated to? The parent?

File: CefSharp.Core/Internals/ClientAdapter.cpp [hunk 13]
```diff
@@ -72,19 +72,42 @@ namespace CefSharp
...
             {
                 return false;
             }
 
-            return handler->OnBeforePopup(_browserControl, gcnew CefFrameWrapper(frame, _browserControl->BrowserAdapter), StringUtils::ToClr(target_url),
-                windowInfo.x, windowInfo.y, windowInfo.width, windowInfo.height, *no_javascript_access);
+            bool returnVal = handler->OnBeforePopup(_browserControl, gcnew CefFrameWrapper(frame, _browserControl->BrowserAdapter), StringUtils::ToClr(target_url),
+				windowInfo.x, windowInfo.y, windowInfo.width, windowInfo.height, *no_javascript_access, newBrowser);
+
+			IWebBrowserInternal^ newBrowserInternal = dynamic_cast<IWebBrowserInternal^>(newBrowser);
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Without looking at this too closely, it appears you've cast more times than is required. What's stopping you from directly casting to `IRenderWebBrowser`?

Should probably check for `newBrowser` to be null before attempting anything. Also code needs comments to document the behavior.

You may not even need to cast as it doesn't appear your using the result.

**Comment by amaitland** [comment]
Does this work in `WinForms`?

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml (hunk 4 above)
Comment (jamespearce2006) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml]: I completely agree - it's not nice. But I could find no other way of delay-loading the control instance, and conform the the MVVM pattern to some degree. I'm not really happy with how the ChromiumWebBrowser gets created first, then added to the visual tree later - it seems the underlying control should probably manage that.

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 6 above)
Comment (jamespearce2006) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: It seems passing null would crash LoadURL()

File: CefSharp.Core/Internals/ClientAdapter.cpp (hunk 12 above)
Comment (jamespearce2006) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: The events are raised against the created instance (the popup). As it's now being displayed as windowless, we can't treat it as a popup.

File: CefSharp.Core/Internals/ClientAdapter.cpp (hunk 13 above)
Comment (jamespearce2006) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Probably - it's been a long while since I've worked in C++, and managed C++ never.

**Comment by jamespearce2006** [comment]
It doesn't not work in Winforms...
I've not broken the Winforms example, but I've not specifically tested the offscreen popup feature... 
I'm selfishly concentrating my efforts on WPF at the moment.

Like I said though, I really just wanted this to be a starting point...

File: CefSharp.Wpf/ChromiumWebBrowser.cs (hunk 6 above)
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Fair enough then :+1:

**Comment by amaitland** [comment]
> Like I said though, I really just wanted this to be a starting point...

And that's fine. It will need to support `WinForms` before merge.

I think to keep the `Example` changes to a minimum, it's probably easiest to spawn a new `Window` that hosts the new `ChromiumWebBrowser` instance. Enough to demo the feature whilst keeping the `Example` simple.

File: CefSharp.Core/Internals/ClientAdapter.cpp (hunk 12 above)
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Makes sense, add comments where it seems logical.

File: CefSharp.Core/Internals/ClientAdapter.cpp (hunk 13 above)
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Check for `nullptr` and use the `IsOffScreen` check where appropriate, most likely remove two casts.

File: CefSharp.Core/Internals/ClientAdapter.cpp [hunk 14]
```diff
@@ -72,19 +72,42 @@ namespace CefSharp
...
+            IWebBrowser^ newBrowser = nullptr;
 
             if (handler == nullptr)
             {
                 return false;
             }
 
-            return handler->OnBeforePopup(_browserControl, gcnew CefFrameWrapper(frame, _browserControl->BrowserAdapter), StringUtils::ToClr(target_url),
-                windowInfo.x, windowInfo.y, windowInfo.width, windowInfo.height, *no_javascript_access);
+            bool returnVal = handler->OnBeforePopup(_browserControl, gcnew CefFrameWrapper(frame, _browserControl->BrowserAdapter), StringUtils::ToClr(target_url),
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: if `returnValue` is true then then opening the popup will be cancelled, so only attempt to assign browser if `false`

**Comment by rassilon** [comment]
Oh hrm.... @amaitland , this interacts in a subtle way with the move to expose IBrowser and IFrame as well as the changes we discussed about adding the IPopupHandler interface.

Would it be too weird for the offscreen popup browser control (i.e. `IWebBrowser`) to only call the `IPopupHandler` methods?

Thoughts?

File: CefSharp.Core/Internals/ClientAdapter.h [hunk 15]
```diff
@@ -62,6 +62,8 @@ namespace CefSharp
             void CloseDevTools();
             void CloseAllPopups(bool forceClose);
 
+            virtual bool IsOffscreen() { return false; }
```
Comment (rassilon) [inline CefSharp.Core/Internals/ClientAdapter.h]: This name is  slightly misleading. Shouldn't it rather be something like `IsPopupInSeparateWebBrowserControl` or something? The parameter documentation for `newBrowser` just says its in a different control, `newBrowser` doesn't need to necessarily be an Offscreen control.

**Comment by rassilon** [comment]
@jamespearce2006 , thank you very much for the launching off point for this issue by creating this starting point for everybody.

**Comment by amaitland** [comment]
> Oh hrm.... @amaitland , this interacts in a subtle way with the move to expose IBrowser and IFrame as well as the changes we discussed about adding the IPopupHandler interface.

Unfortunately :frowning: I was hopeful that we might be able to tell `CEF` that it's not actually a popup, though haven't looked into that yet.

> Would it be too weird for the offscreen popup browser control (i.e. IWebBrowser) to only call the IPopupHandler methods?

Can you elaborate a little more?

**Comment by rassilon** [comment]
> > Would it be too weird for the offscreen popup browser control (i.e. IWebBrowser) to only call the IPopupHandler methods?

> Can you elaborate a little more?

Well, since it's a popup window, IWebBorwser wouldn't call the event handlers that only apply to  MainFrame browsers.

**Comment by amaitland** [comment]
> Well, since it's a popup window, IWebBorwser wouldn't call the event handlers that only apply to MainFrame browsers

So the `Dependency Properties` wouldn't work anymore?

**Comment by amaitland** [comment]
(From memory there are a few bound to the `CEF` events in question).

**Comment by amaitland** [comment]
On a side note, we should look at exposing `CefPopupFeatures` and `CefBrowserSettings` when opening a popup.

**Comment by rassilon** [comment]
> So the Dependency Properties wouldn't work anymore?

Not unless we create a special check for this case. (Which is what I think this PR's IsOffscreen`method is doing.

**Comment by amaitland** [comment]
I think the combination of `IsPopup` and `IsOffScreen` combined should be enough to differentiate.

**Comment by amaitland** [comment]
Probably need to conditionally call `CefBrowserHost::CreateBrowser` now, doubt it will be needed for this popup scenario.

File: CefSharp.Core/Internals/ClientAdapter.cpp [hunk 16]
```diff
@@ -72,19 +72,42 @@ namespace CefSharp
...
+                windowInfo.x, windowInfo.y, windowInfo.width, windowInfo.height, *no_javascript_access, newBrowser);
+
+            IWebBrowserInternal^ newBrowserInternal = dynamic_cast<IWebBrowserInternal^>(newBrowser);
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Should really check `newBrowser != null` first up, then only proceed of that's the cast.

**Comment by amaitland** [comment]
@rassilon I'm somewhat inclined to resolve the `IPopupHandler/whatever we're going to call it` changes first then look at merging this once there's a clearer picture of how that will work. Thoughts?

File: CefSharp.Wpf.Example/Views/BrowserTabView.xaml (hunk 4 above)
Comment (amaitland) [inline CefSharp.Wpf.Example/Views/BrowserTabView.xaml]: > it seems the underlying control should probably manage that.

Which control were you thinking? I'm sure we can come up with something, just have to ponder for a bit....

**Comment by jamespearce2006** [comment]
I've simplified the XAML example, as requested, and made it simply open the popup in a new window, rather than trying to shoe-horn into the existing MVVM pattern.

**Comment by jamespearce2006** [comment]
I had a look at simplifying the casts in ClientAdapter::OnBeforePopup, but I'm not sure I can. I can't just check the IsOffscreen property because we're dealing with a newly created client, not the current instance. 

The first cast is to get the IWebBrowserInternal, to ensure it's something that belongs to us. This will work whether it's offscreen or not.
The next cast is to do some specific work for IRenderWebBrowser (set windowInfo set windowless).

It's quite likely I'm misunderstanding something, though.

**Comment by amaitland** [comment]
> I've simplified the XAML example, as requested, and made it simply open the popup in a new window, rather than trying to shoe-horn into the existing MVVM pattern.

Nice one :+1: 

> I had a look at simplifying the casts in ClientAdapter::OnBeforePopup, but I'm not sure I can. I can't just check the IsOffscreen property because we're dealing with a newly created client, not the current instance.

If the current instance is `OffScreen` then the newly created instance should also be `OffScreen`? We will need to add a separate property as @rassilon suggested as the `IsOffScreen()` checks you added won't work for `WinForms`. Rather than littering the place with `if else` blocks we might need to consider another layer of abstraction, possibly a plug-able implementation to handle different scenarios, some more though is required I think (It maybe overkill).

**Comment by rassilon** [comment]
Yeah, I agree we should merge this after the IPopupHandler changes.

**Comment by jamespearce2006** [comment]
> If the current instance is OffScreen then the newly created instance should also be OffScreen?

I wouldn't have thought that's guaranteed. Presumably there's nothing to stop a windowed instance creating an offscreen popup.

**Comment by arsher** [comment]
Hey guys,

I've checked this out because I'm going to need something like this soon. But I found a caveat with the current approach.
When you're using window.open and staying in the same domain chromium can put the child window into the same process and also gives you access to the parent window (and from the parent to the children) through window.opener so the js context is accessible and you can also pass dom elements. But this solution has a new CefClient for each control so this pretty neat feature won't work. I did some experiments with reusing the ClientAdapter and creating a CefRenderHandler for each child window and it seems to be working, but it would require some thinking on the architecture to fit this in with proper cleanup etc...What do you think? Is it worth it to pursuing this?

David

**Comment by amaitland** [comment]
#1150 was merged into `master`, so closing this.

**Pull request closed without merging**